### ISSUE TITLE
High object allocations when using `writeString` and `writeBytes` from a different thread.

### ISSUE THREAD

**Issue by NiteshKant** [ReactiveX/RxNetty#395]
`HttpServerResponse` and `ObservableConnection` provide methods like `writeString()`, `writeBytes()`, `write(R r, ContentTransformer<T, R> transformer)` and their variants with `*AndFlush()` that uses netty's `ByteBufAllocator` to allocate new `ByteBuf` if required.

RxNetty defaults to using a `PooledByteBufAllocator` that has a **thread local pool** of `ByteBuf`s. In case where this allocator is used to allocate new buffers, from a large pool of threads (eg: `Schedulers.io()` with high concurrency), this will create `ByteBuf` pools on each of those threads. This will cause a memory bloat.
##### Reproducer

(**For this test, I had turned off off-heap buffers for easier insight using mission control, the property is `-Dio.netty.noPreferDirect=true`**)

I did a small test with the following two server variants:

This writes to the response from a different thread (`Schedulers.io()` in `Observable.interval()`)

``` java
        return RxNetty.createHttpServer(8090, (request, response) -> {
            return Observable.interval(1, TimeUnit.MILLISECONDS, Schedulers.io())
                             .take(20)
                             .flatMap(aTick -> response.writeStringAndFlush("Hello!!!"));
        });
```

and this writes in the subscribing thread (eventloop)

``` java
        return RxNetty.createHttpServer(8091, (request, response) -> {
            return Observable.range(1, 20)
                             .flatMap(aTick -> response.writeStringAndFlush("Hello!!!"));
        });
```

On running the below client on both the servers:

``` java

        final CountDownLatch latch = new CountDownLatch(100);

        for (int i = 0; i < 100; i++) {
            RxNetty.<ByteBuf, ByteBuf>newHttpClientBuilder("localhost", 8091)
                   .pipelineConfigurator(new HttpClientPipelineConfigurator<>())
                   .build()
                   .submit(HttpClientRequest.createGet("/"))
                   .flatMap(response -> response.getContent()
                                                .<String>map(content -> content.toString(Charset.defaultCharset())))
                   .finallyDo(latch::countDown)
                   .forEach(System.out::println);
        }

        latch.await();
```

I found the below object allocations:
###### Write in eventloop

![image](https://cloud.githubusercontent.com/assets/820005/8813915/d9e70698-2fbb-11e5-94ec-2a54f603f714.png)

![image](https://cloud.githubusercontent.com/assets/820005/8813982/8ce9e4ea-2fbc-11e5-9ba2-99e2ed52a04a.png)
###### Write in `Schedulers.io()`

![image](https://cloud.githubusercontent.com/assets/820005/8813936/201b4980-2fbc-11e5-983d-8e4ceeb7512c.png)

![image](https://cloud.githubusercontent.com/assets/820005/8813976/7cfdae86-2fbc-11e5-8110-da2873e99acd.png)

Looking at the above, it is evident that there are way too many allocations of `PoolThreadCache$MemoryRegionCache$Entry` which is an entry in the per-thread cache of `byte[]`, in this case.

**Comment by NiteshKant** [comment]
This is not the case with 0.5.x as it does not do these allocations on the caller thread (writer) but does the allocation within netty's handler which always runs on an eventloop. The fix would be to backport the approach from 0.5.x

**Comment by NiteshKant** [comment]
The allocations after applying the changes in PR #396 are as follows:

###### Write in Schedulers.io()

![image](https://cloud.githubusercontent.com/assets/820005/8835167/96f3bc70-306d-11e5-85bc-ec665ec2195f.png)

###### Write in eventloop

![image](https://cloud.githubusercontent.com/assets/820005/8835153/84eff908-306d-11e5-9943-5eb0ca95b1c0.png)
